This the homework for week 1 of the MLOps zoomcamp

In [70]:
!python -V
!pip install seaborn
!pip install scikit-learn
!pip install pickle

Python 3.8.12

[notice] A new release of pip is available: 23.0 -> 23.1.2
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.0 -> 23.1.2
[notice] To update, run: pip install --upgrade pip
ERROR: Could not find a version that satisfies the requirement pickle (from versions: none)
ERROR: No matching distribution found for pickle

[notice] A new release of pip is available: 23.0 -> 23.1.2
[notice] To update, run: pip install --upgrade pip


In [69]:
import pandas as pd
import numpy as np
import pyarrow
import pickle
import os
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression, LogisticRegression, Lasso, Ridge
from sklearn.metrics import mean_squared_error


In [71]:
def read_dataframe(path):
    df = pd.read_parquet(path)
    df['duration'] = df.tpep_dropoff_datetime - df.tpep_pickup_datetime
    df['duration'] = df.duration.dt.total_seconds()/60
    categorical = ['PULocationID','DOLocationID']
    
    df[categorical] = df[categorical].astype(str)
    
    return df
    

In [72]:
path = 'data/yellow_tripdata_2022-01.parquet'
df = pd.read_parquet(path)
df
# len(df.columns)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
2463926         2  2022-01-31 23:36:53   2022-01-31 23:42:51              NaN   
2463927         2  2022-01-31 23:44:22   2022-01-31 23:55:01              NaN   
2463928         2  2022-01-31 23:39:00   2022-01-31 23:50:00              NaN   
2463929         2  2022-01-31 23:36:42   2022-01-31 23:48:45              NaN   
2463930         2  2022-01-31 23:46:00   2022-02-01 00:13:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 3.80         1.0                  N           142   
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
2463926           1.32         NaN               None            90   
2463927           4.19         NaN               None           107   
2463928           2.10         NaN               None           113   
2463929           2.92         NaN               None           148   
2463930           8.94         NaN               None           186   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 236             1        14.50    3.0      0.5        3.65   
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
2463926           170             0         8.00    0.0      0.5        2.39   
2463927            75             0        16.80    0.0      0.5        4.35   
2463928           246             0        11.22    0.0      0.5        2.00   
2463929           164             0        12.40    0.0      0.5        0.00   
2463930           181             0        25.48    0.0      0.5        6.28   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         21.95   
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
2463926           0.0                    0.3         13.69   
2463927           0.0                    0.3         24.45   
2463928           0.0                    0.3         16.52   
2463929           0.0                    0.3         15.70   
2463930           0.0                    0.3         35.06   

         congestion_surcharge  airport_fee  
0                         2.5          0.0  
1                         0.0          0.0  
2                         0.0          0.0  
3                         2.5          0.0  
4                         2.5          0.0  
...                       ...          ...  
2463926                   NaN          NaN  
2463927                   NaN          NaN  
2463928                   NaN          NaN  
2463929                   N

Q_1: Read the data for January. How many columns are there?
Answer: 19

In [73]:
df_train = read_dataframe('data/yellow_tripdata_2022-01.parquet')
df_val = read_dataframe('data/yellow_tripdata_2022-02.parquet')

In [74]:
len(df_train), len(df_val)

df_train

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
2463926         2  2022-01-31 23:36:53   2022-01-31 23:42:51              NaN   
2463927         2  2022-01-31 23:44:22   2022-01-31 23:55:01              NaN   
2463928         2  2022-01-31 23:39:00   2022-01-31 23:50:00              NaN   
2463929         2  2022-01-31 23:36:42   2022-01-31 23:48:45              NaN   
2463930         2  2022-01-31 23:46:00   2022-02-01 00:13:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag PULocationID  \
0                 3.80         1.0                  N          142   
1                 2.10         1.0                  N          236   
2                 0.97         1.0                  N          166   
3                 1.09         1.0                  N          114   
4                 4.30         1.0                  N           68   
...                ...         ...                ...          ...   
2463926           1.32         NaN               None           90   
2463927           4.19         NaN               None          107   
2463928           2.10         NaN               None          113   
2463929           2.92         NaN               None          148   
2463930           8.94         NaN               None          186   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                236             1        14.50    3.0      0.5        3.65   
1                 42             1         8.00    0.5      0.5        4.00   
2                166             1         7.50    0.5      0.5        1.76   
3                 68             2         8.00    0.5      0.5        0.00   
4                163             1        23.50    0.5      0.5        3.00   
...              ...           ...          ...    ...      ...         ...   
2463926          170             0         8.00    0.0      0.5        2.39   
2463927           75             0        16.80    0.0      0.5        4.35   
2463928          246             0        11.22    0.0      0.5        2.00   
2463929          164             0        12.40    0.0      0.5        0.00   
2463930          181             0        25.48    0.0      0.5        6.28   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         21.95   
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
2463926           0.0                    0.3         13.69   
2463927           0.0                    0.3         24.45   
2463928           0.0                    0.3         16.52   
2463929           0.0                    0.3         15.70   
2463930           0.0                    0.3         35.06   

         congestion_surcharge  airport_fee   duration  
0                         2.5          0.0  17.816667  
1                         0.0          0.0   8.400000  
2                         0.0          0.0   8.966667  
3                         2.5          0.0  10.033333  
4                         2.5          0.0  37.533333  
...                       ...          ...        ...  
2463926                   NaN          NaN   5.966667  
2463927                   NaN          NaN  10.650000

In [75]:
dur = df_train['duration']

np.std(dur)

46.445295712725304

Q_2: What's the standard deviation of the trips duration in January?
Answer: 46.45

In [76]:
before = len(df_train)
df_train = df_train[(df_train.duration > 1) & (df_train.duration <= 60)]
df_val = df_val[(df_val.duration > 1) & (df_val.duration <= 60)]
after = len(df_train)

fraction = after / before

fraction




0.9826216724413143

Q_3: What fraction of the records left after you dropped the outliers?
Answer: 98%

In [77]:
categorical = ['PULocationID','DOLocationID']
numerical = ['trip_distance']

dv = DictVectorizer()

train_dicts = df_train[categorical + numerical].to_dict(orient='records')
X_train = dv.fit_transform(train_dicts)

val_dicts = df_val[categorical + numerical].to_dict(orient='records')
X_val = dv.transform(val_dicts)

In [82]:
X_train.ndim


<2917848x516 sparse matrix of type '<class 'numpy.float64'>'
	with 8753538 stored elements in Compressed Sparse Row format>

Q_4: What's the dimensionality of this matrix (number of columns)?
Answer: 2

In [83]:
target = 'duration'
y_train = df_train[target].values
y_val = df_val[target].values

In [86]:
lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred = lr.predict(X_train)

mean_squared_error(y_train, y_pred, squared=False)

7.000035096658634

Q_5: What's the RMSE on train?
Answer: 6.99 or 7.00

In [87]:
lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred = lr.predict(X_val)

mean_squared_error(y_val, y_pred, squared=False)

7.794205733165656

Q_5: What's the RMSE on train?
Answer: 7.79